Windows: Build and use gotestsum for running all tests

thaJeztah opened the pull request:
carry of https://github.com/moby/moby/pull/39971 to make it pick-up the Jenkinsfile changes

thaJeztah commented on Jenkinsfile:
```diff
@@ -828,6 +828,7 @@ pipeline {
                     }
                     post {
                         always {
+                            junit testResults: 'bundles/junit-report.xml', allowEmptyResults: true
```
I can disable the `allowEmptyResults` for now, so that it's clear if it failed

Perhaps we should do a `dir` (e.g.) too see what files are present, and if the location in which it's stored is correct

thaJeztah commented:
one failure, which is known to be flaky; https://github.com/moby/moby/issues/32219

https://ci.docker.com/public/job/moby/job/PR-39998/1/execution/node/209/log/?consoleFull

```
01:16:22.548      --- FAIL: TestDockerSuite/TestRestartContainerwithRestartPolicy (9.00s)
01:16:22.548          cli.go:29: assertion failed: 
01:16:22.548              Command:  d:\CI\PR-39998\1\binary\docker.exe restart 84a49937203b5fa980e4a6b69ab5af98e8d363e1862ccb9caa49da62015006a5
01:16:22.548              ExitCode: 1
01:16:22.548              Error:    exit status 1
01:16:22.548              Stdout:   
01:16:22.548              Stderr:   Error response from daemon: Cannot restart container 84a49937203b5fa980e4a6b69ab5af98e8d363e1862ccb9caa49da62015006a5: Container 84a49937203b5fa980e4a6b69ab5af98e8d363e1862ccb9caa49da62015006a5 is not running
01:16:22.548              
01:16:22.548              
01:16:22.548              Failures:
01:16:22.548              ExitCode was 1 expected 0
01:16:22.548              Expected no error
```

It did run the archive junit results step; https://ci.docker.com/public/job/moby/job/PR-39998/1/execution/node/355/log/

But I don't see the test-results for windows;
https://ci.docker.com/public/job/moby/job/PR-39998/1/testReport/

andrewhsu commented:
To help with development iteration, you could throw in a temporary git commit to disable tests except for `windowsRS5`.
https://github.com/moby/moby/blob/5b57f41241cb0d8df5886a1db3d7920494a9e432/Jenkinsfile#L11-L17

Commit:
Build and use gotestsum for running all tests

1. Dockerfile.Windows modified to build gotestsum.exe

2. Use gotestsum.exe in invoking the execution of:

   (a) Unit tests (run in containers),
   (b) Integration tests (run outside containers)
   (c) Integration-cli (run outside containers)

No changes made to other categories of tests (e.g.
LCOW).

3. Copy .xml files produced by gotestsum in
   appropriate paths where Jenkins can ingest them

4. Modify Jenkinsfile to mark results output as
   being jUnit "type" as well as to archive the
   .xml test result files as artifacts.

Signed-off-by: Vikram bir Singh <vikrambir.singh@docker.com>
Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented:
Looks like it still doesn't write results in the right location;

https://ci.docker.com/public/job/moby/job/PR-39998/5/execution/node/356/log/


```
01:01:32.396  Recording test results
01:01:34.755  None of the test reports contained any result
```

thaJeztah commented:
@vikramhh I rebased your branch again; I think your local git was not up-to-date with the upstreams.

thaJeztah commented on Dockerfile.windows:
```diff
@@ -250,18 +256,89 @@ RUN `
-  New-Item -ItemType Directory -Path ${GOPATH}\src\github.com\docker\docker | Out-Null; `
+  #TODO - Per https://docs.docker.com/engine/reference/builder/ 
+  #       the below line is redundant and should be removed.
+  #New-Item -ItemType Directory -Path ${GOPATH}\src\github.com\docker\docker | Out-Null; `
```
We should update those docs a bit. Although `WORKDIR` _will_ create the directory if it doesn't exist, when it _does_ it always creates the directory as `root` (or likely the equivalent on Windows).

This means that if the Dockerfile is using a different `USER` (either during build, or when running the container, so with `--user=<some-user>:<some-group>`, hence in the following:

```Dockerfile
FROM foo
USER some_non_privileged_user
WORKDIR /some/path/
```

`/some/path/` will be owned by `root`, and not `some_non_privileged_user`

For that reason, it's often good practice to create the directory manually, instead of relying on the side effect of `WORKDIR` creating the directory.

vikramhh replied:
Changed to create the directory explicitly and removed the WORKDIR directive

thaJeztah commented on Dockerfile.windows:
```diff
@@ -250,18 +256,89 @@ RUN `
+  C:\git\cmd\git config --global core.autocrlf true;
+
+# Ensure all directories exist that we will require below....
+WORKDIR ${GOPATH}\${DOCKER_SRC_REL_LOCATION}\${BUNDLES_REL_LOCATION}
```
See my comment above, I don't think we should abuse `WORKDIR`'s  side-effect as a replacement for `mkdir`

vikramhh replied:
Removed as it was there for its side-effect alone

thaJeztah commented on Dockerfile.windows:
```diff
@@ -250,18 +256,89 @@ RUN `
+  } `
   `
-  Write-Host INFO: Completed
+  Write-Host INFO: Building gotestsum...; `
```
Should we move this inside the `Build-GoTestSum` function?

vikramhh replied:
I do not know a compelling argument either way. Why would you choose one over the other?

vikramhh replied:
Done

thaJeztah commented on Dockerfile.windows:
```diff
@@ -250,18 +256,89 @@ RUN `
+      } `
+    } `
+  } `
+  Function Build-GoTestSum() { `
```
The scripts that we're writing in the Dockerfile now are a bit lengthy, and more complex to maintain (having to keep all the line-continuations (```) into account.

Perhaps we should put these function in a `.ps` file (easier to write/maintain), and;

Somewhere at the top of the Dockerfile (assuming the functions themself don't change often);

```Dockerfile
COPY hack/dockerfile/install/gotestsum.installer.ps1 hack/dockerfile/install/gotestsum.installer.ps1
```

And later on, call those functions


```Dockerfile
RUN . "hack/dockerfile/install/gotestsum.installer.ps1" Build-GoTestSum 
```

(or however that should be done in PowerShell)

vikramhh replied:
Good suggestion and something I have put on my plate. It will be done in a separate PR

thaJeztah commented on Jenkinsfile:
```diff
@@ -8,12 +8,12 @@ pipeline {
-        booleanParam(name: 'windowsRS1', defaultValue: false, description: 'Windows 2016 (RS1) Build/Test')
+        booleanParam(name: 'unit_validate', defaultValue: false, description: 'amd64 (x86_64) unit tests and vendor check')
+        booleanParam(name: 'amd64', defaultValue: false, description: 'amd64 (x86_64) Build/Test')
+        booleanParam(name: 'arm64', defaultValue: false, description: 'ARM (arm64) Build/Test')
```
These changes should be removed now as they were only for testing

vikramhh replied:
Done

vikramhh replied:
Changed all except `windowsRS1` which should be kept `true` by default to ensure that before merge, we verify that RS1 does not break. After that, we could change it back.

thaJeztah commented on hack/ci/windows.ps1:
```diff
@@ -864,10 +903,13 @@ Try {
                     $env:INTEGRATION_TESTFLAGS = "-test.v"
                 }
-                Set-Location "$env:SOURCES_DRIVE`:\$env:SOURCES_SUBDIR\src\github.com\docker\docker"
+                #Set-Location "$env:SOURCES_DRIVE`:\$env:SOURCES_SUBDIR\src\github.com\docker\docker"
```
Why was this commented-out? Is it no longer needed? (if so, we should remove it)

vikramhh replied:
Done

thaJeztah commented on hack/make.ps1:
```diff
@@ -94,6 +94,8 @@ param(
 $ErrorActionPreference = "Stop"
 $ProgressPreference = "SilentlyContinue"
 $pushed=$False  # To restore the directory if we have temporarily pushed to one.
+Set-Variable GOTESTSUM_COMMIT -option Constant -value 'v0.3.5'
```
Is this variable still used? (I see it's also declared in the Dockerfile)?

vikramhh replied:
Removed

thaJeztah commented on hack/make.ps1:
```diff
@@ -319,36 +321,58 @@ Function Run-UnitTests() {
+    $env:DOCKER_INTEGRATION_DAEMON_DEST = $bundlesDir + "\tmp"
     $dirs = go list -test -f '{{- if ne .ForTest `"`" -}}{{- .Dir -}}{{- end -}}' .\integration\...
-    $integration_api_dirs = @()
+    #$integration_api_dirs = @()
```
If this is no longer needed, we should remove it, instead of commenting-out

vikramhh replied:
Done

thaJeztah commented on hack/make.ps1:
```diff
@@ -319,36 +321,58 @@ Function Run-UnitTests() {
-    ForEach($dir in $integration_api_dirs) {
+        #$integration_api_dirs += $dir
+        #Write-Host "Building test suite binary $dir"
+        #go test -c -o "$dir\test.exe" $dir
```
Same here.


Wondering though why we're no longer building the binaries here? (perhaps you can explain why this changed)

vikramhh replied:
Instead of building the test binary and then running it, we are just running the test. Why we need to do it in two different stages?

thaJeztah commented on hack/make.ps1:
```diff
@@ -319,36 +321,58 @@ Function Run-UnitTests() {
         Write-Host "Running $($PWD.Path)"
         $pinfo = New-Object System.Diagnostics.ProcessStartInfo
-        $pinfo.FileName = "$($PWD.Path)\test.exe"
+        #gotestsum.exe is on the path
```
Should we be explicit here and use `"$($PWD.Path)\gotestsum.exe"` ? 

Otherwise, I think we can remove the comment

vikramhh replied:
It is located elsewhere and is explicitly added to the PATH [see line 899 in `windows.ps1`]. So using `"$($PWD.Path)\gotestsum.exe"` may not be right. 

I have removed the comment.

thaJeztah commented on hack/make.ps1:
```diff
@@ -319,36 +321,58 @@ Function Run-UnitTests() {
         $pinfo.UseShellExecute = $false
-        $pinfo.Arguments = $env:INTEGRATION_TESTFLAGS
+        $tmpArgs = "--format=standard-quiet --jsonfile=$jsonFilePath --junitfile=$xmlFilePath -- $env:INTEGRATION_TESTFLAGS"
+        #Write-Host $tmpArgs
```
Looks like this was there for debugging; do we need the intermediate `$tmpArgs` variable? Or can we set it directly below;

```powershell
$pinfo.Arguments = "--format=standard-quiet --jsonfile=$jsonFilePath --junitfile=$xmlFilePath -- $env:INTEGRATION_TESTFLAGS"
```

vikramhh replied:
Changed

thaJeztah commented on hack/make.ps1:
```diff
@@ -319,36 +321,58 @@ Function Run-UnitTests() {
         $p.Start() | Out-Null
         $p.WaitForExit()
         $err = $p.StandardError.ReadToEnd()
+        Write-Host "$err"
```
I _think_ this is to print `stderr` output if there's no error that happened, correct?

In case of an error (`$LASTEXITCODE -ne 0`), this will now be printed _twice_. Think it should be changed to either;

```powershell
if (($LASTEXITCODE -ne 0) -and ($err -notlike "*warning: no tests to run*")) {
    Throw "Integration tests failed: $err"
} else {
    Write-Host "$err"
}
```

Or (if `Throw` exits the script (not sure if if does so?)); perhaps add a comment as well to explain why we're printing `$err`

```powershell
if (($LASTEXITCODE -ne 0) -and ($err -notlike "*warning: no tests to run*")) {
    Throw "Integration tests failed: $err"
}

# Write stderr output
Write-Host "$err"
```

andrewhsu commented:
I'm starting to see the junit reports of the test results in jenkins UI now, which is a great step: https://ci.docker.com/public/blue/organizations/jenkins/moby/detail/PR-39998/38/tests

andrewhsu commented:
Workaround for now: #40155

thaJeztah commented on Dockerfile.windows:
```diff
@@ -249,19 +255,87 @@ RUN `
-  C:\git\cmd\git config --global core.autocrlf true; `
+  C:\git\cmd\git config --global core.autocrlf true;
+
+WORKDIR ${GOPATH}
```
Was there a reason to change to this location here? Looks like the script takes care to going to the correct directory, so changing the `WORKDIR` here, only to change it back to `$GOPATH\src\github.com\docker\docker` after, may not be needed.

Effectively, now we do;

```Dockerfile
WORKDIR ${GOPATH}
RUN Build-GoTestSum
WORKDIR ${GOPATH}\src\docker\docker
```

vikramhh replied:
Moved the change of directory to `Env:GOPATH` explicitly in the script and removed the first `WORKDIR` above.

thaJeztah commented on Dockerfile.windows:
```diff
@@ -249,19 +255,87 @@ RUN `
 # Copy the sources into the container
 COPY . .
+
+RUN `
```
This looks redundant; `docker build` will already show that the build completed

vikramhh replied:
Removed

thaJeztah commented on hack/make.ps1:
```diff
@@ -94,6 +94,7 @@ param(
 $ErrorActionPreference = "Stop"
 $ProgressPreference = "SilentlyContinue"
 $pushed=$False  # To restore the directory if we have temporarily pushed to one.
+Set-Variable GOTESTSUM_LOCATION -option Constant -value "$env:GOPATH/src/gotest.tools/gotestsum"
```
If we compile `gotestsum` to be put in `$GOBIN` (which defaults to `$GOPATH\bin`), we don't need this variable, and we can just call `gotestsum.exe ...`

If would mean we have to keep track of less things, and can simplify the scripts

vikramhh replied:
Actually this will require a separate copy to $GOBIN and introducing or at least using a new variable $GOBIN which is currently not being used anywhere in these scripts. And it will save us from having to track things in only one place. So I am afraid it will actually end up increasing complexity rather than decreasing it.

thaJeztah commented on hack/make.ps1:
```diff
@@ -363,6 +379,10 @@ Try {
 
+    # Create the bundles directory if it doesn't exist
+    $bundlesDir = $root + "\bundles"
+    if (-not (Test-Path $bundlesDir)) { New-Item $bundlesDir -ItemType Directory | Out-Null }
```
Can we change to the equivalent of `mkdir -p` here? Then we can also get rid of the intermediate variable

I think that's something like;

```powershell
New-Item -Force "$root\bundles" -ItemType Directory | Out-Null
```


Actually, I see the `Push-Location` above, so we could probably even use a relative path (I see that was was used previously)

```powershell
New-Item -Force ".\bundles" -ItemType Directory | Out-Null
```

vikramhh replied:
I did make the change to the latter by making it equivalent to "mkdir -p". However the intermediate variable is then used at other places as well. Making it an explicit value instead of the variable will require us to still set this variable in other places for those usages [or convert those in multiple places to use explicit paths]. But more important, I am afraid it will mask the intent [that it is the same directory which is being referred to in all those disparate locations] and make the code hard to maintain. So I would argue for leaving this as-is.

thaJeztah reviewed (commented):
Thanks! Left my comments inline

andrewhsu commented:
Talked with vikram about this PR today. Looks like the junit report of failures is enough to make sure the github status of the PR checks is red even though the unix return code of that step is 0. This seems acceptable to get this PR in but needs work later to get the unix return code of the step to accurately reflect test failures.

thaJeztah commented on Dockerfile.windows:
```diff
@@ -249,19 +286,21 @@ RUN `
+  # Ensure all directories exist that we will require below....
+  $srcDir = """$Env:GOPATH`\src\github.com\docker\docker\bundles"""; `
+  Write-Host INFO: Ensuring existence of directory $srcDir...; `
+  New-Item -ItemType Directory -Path $srcDir | Out-Null; `
```
We should probably add `-Force` here (to get the equivalent of `mkdir -p`); we don't want this to fail if the `bundles` directory exists.

vikramhh replied:
Done

thaJeztah commented on Dockerfile.windows:
```diff
@@ -249,19 +286,21 @@ RUN `
-  New-Item -ItemType Directory -Path ${GOPATH}\src\github.com\docker\docker | Out-Null; `
+  # Ensure all directories exist that we will require below....
+  $srcDir = """$Env:GOPATH`\src\github.com\docker\docker\bundles"""; `
+  Write-Host INFO: Ensuring existence of directory $srcDir...; `
```
does running the command below itself show up in the logs? If so, this might be redundant

vikramhh replied:
This is primarily for consistency with the code around this line. Dockerfile.Windows uses this patttern consistently [look at lines above and below this piece of code].

thaJeztah commented on Dockerfile.windows:
```diff
@@ -208,6 +211,40 @@ RUN `
+    $GotestsumBinPath=Join-Path -Path $GotestsumPath -ChildPath "gotestsum.exe"; `
+    Write-Host "INFO: Building gotestsum version $Env:GOTESTSUM_COMMIT in $Env:GOPATH" `
+    $optsForGet = @('"get"', '"-d"', '"gotest.tools/gotestsum"'); ` 
+    $logFileName = 'gotestsum.get.out';  `
```
This variable is no longer used so can be removed

vikramhh replied:
Done - removed all references to log file names.

thaJeztah commented on Dockerfile.windows:
```diff
@@ -208,6 +211,40 @@ RUN `
+    Write-Host "INFO:     Sources obtained for gotestsum..."; `
+    Push-Location $GotestsumPath; `
+    $optsForCheckout = @('"checkout"', '"-q"', """$Env:GOTESTSUM_COMMIT"""); ` 
+    $logFileName = 'gotestsum.git.out';  `
```
This variable is no longer used so can be removed

vikramhh replied:
Done

thaJeztah commented on Dockerfile.windows:
```diff
@@ -208,6 +211,40 @@ RUN `
+    $savedExitCode = $LASTEXITCODE; `
+    if ($savedExitCode -eq 0) { `
+      Write-Host "INFO:     Checkout done for gotestsum..."; `
+      $logFileName = 'gotestsum.build.out';  `
```
This variable is no longer used so can be removed

vikramhh replied:
Done

thaJeztah commented on Dockerfile.windows:
```diff
@@ -208,6 +211,40 @@ RUN `
+    Push-Location $GotestsumPath; `
+    $optsForCheckout = @('"checkout"', '"-q"', """$Env:GOTESTSUM_COMMIT"""); ` 
+    $logFileName = 'gotestsum.git.out';  `
+    &git $optsForCheckout; `
```
See my previous review; could those options be moved inline here?

vikramhh replied:
I feel that the above is much cleaner and easier to read. I am open to changing it if there is a reason why we should inline the arguments here.

thaJeztah replied:
Perhaps personally, but in order to understand what this line does, I have look for where `$optsForCheckout` is declared; so read up two lines higher, mentally "remove" the additional quoting that's needed there, and "reconstruct" that this says;

```console
git checkout -q "$GOTESTSUM_COMMIT"
```

thaJeztah commented on Dockerfile.windows:
```diff
@@ -208,6 +211,40 @@ RUN `
+      Write-Host "INFO:     Checkout done for gotestsum..."; `
+      $logFileName = 'gotestsum.build.out';  `
+      $optsForBuild = @('"build"', '"-buildmode=exe"'); ` 
+      &go $optsForBuild; `
```
Same here

vikramhh replied:
Same as above - we can change it if you feel strongly for some reason that arguments must be inline only.

thaJeztah reviewed (commented):
I see you updated some parts, thanks! (left some comments inline); looks like it's currently failing (wondering if we need to add something to `$PATH`?)

```

[2019-10-31T23:53:17.936Z] INFO: Ensuring existence of directory C:\gopath\src\github.com\docker\docker\bundles ...
[2019-10-31T23:53:17.936Z] INFO: Configuring git core.autocrlf...
[2019-10-31T23:53:17.936Z] INFO: Building gotestsum version v0.3.5 in C:\gopath  = get -d gotest.tools/gotestsum
[2019-10-31T23:53:20.103Z] & : The term 'go' is not recognized as the name of a cmdlet, function, script 
[2019-10-31T23:53:20.103Z] file, or operable program. Check the spelling of the name, or if a path was 
[2019-10-31T23:53:20.103Z] included, verify that the path is correct and try again.
[2019-10-31T23:53:20.103Z] At line:1 char:1711
[2019-10-31T23:53:20.103Z] + ... st.tools/gotestsum'); $logFileName = 'gotestsum.get.out'; &go $optsFo ...
```

thaJeztah commented on Dockerfile.windows:
```diff
@@ -208,6 +211,40 @@ RUN `
   `
+  Function Build-GoTestSum() { `
+    $GotestsumPath=Join-Path -Path $Env:GOPATH -ChildPath "src\gotest.tools\gotestsum"; `
+    $GotestsumBinPath=Join-Path -Path $GotestsumPath -ChildPath "gotestsum.exe"; `
```
Looks like this one isn't used anymore as well

vikramhh replied:
Removed

thaJeztah commented on hack/ci/windows.ps1:
```diff
@@ -774,11 +786,36 @@ Try {
+            
+            # Saving where jenkins will take a look at.....
+            $bundlesDir = "bundles"
+            if (-not (Test-Path $bundlesDir)) { New-Item $bundlesDir -ItemType Directory | Out-Null }
```
oh; this one could also use `-Force` (equivalent of `mkdir -p`);

```powershell
 New-Item -Force "bundles" -ItemType Directory | Out-Null
```

vikramhh replied:
Done - however instead of explicit use of the value of the directory name, I am choosing to keep the variable because it is used elsewhere too.

thaJeztah commented on hack/make.ps1:
```diff
@@ -363,6 +379,10 @@ Try {
 
+    # Ensure the bundles directory exists
+    $bundlesDir = $root + "\bundles"
+    New-Item -Force $bundlesDir -ItemType Directory | Out-Null
```
Could this use a relative path?

```powershell
New-Item -Force ".\bundles" -ItemType Directory | Out-Null
```

vikramhh replied:
That would mean we will need to set this in other places anyways where we cannot use a relative path[look at the function `Run-IntegrationTests` for example]. And it would also hide the intent that it is exactly the same directory that is being referred to in all of those places. The way it is used now, there are no dependencies that would need to be considered by someone modifying this code in the future. So I suggest we keep it as-is.

thaJeztah replied:
Ah, yes, I see; it's not a "local" variable, but used globally. Wondering if we could somehow set it at the top; is there a convention to distinguish "local" from "global" variables in PowerShell? (I see other variables use an Uppercase) (or perhaps even a `const`?)

vikramhh replied:
This variable is in global scope and one can indeed distinguish locals and globals. Powershell does have constants but one cannot make a value constant after definition. One can however make it "Read Only" anytime - which is what I did.

thaJeztah commented on Dockerfile.windows:
```diff
@@ -208,6 +211,40 @@ RUN `
+    if ($savedExitCode -eq 0) { `
+      Write-Host "INFO:     Checkout done for gotestsum..."; `
+      $logFileName = 'gotestsum.build.out';  `
+      $optsForBuild = @('"build"', '"-buildmode=exe"'); ` 
```
note: with `go build -o <some path>` you can specify where the binary should be stored, which could make things easier to handle (you can collect the binaries in a standard directory)

vikramhh commented on Dockerfile.windows:
```diff
@@ -174,9 +175,14 @@ ENV GO_VERSION=${GO_VERSION} `
+    FROM_DOCKERFILE=1 `
+    GOTESTSUM_COMMIT=${GOTESTSUM_COMMIT} `
+    GOTESTSUM_LOCATION="src\gotest.tools\gotestsum" `
+    DOCKER_SRC_REL_LOCATION="src\github.com\docker\docker" `
```
Done

vikramhh commented on Dockerfile.windows:
```diff
@@ -249,19 +255,87 @@ RUN `
+    $GotestsumPath=Join-Path -Path $Env:GOPATH -ChildPath $Env:GOTESTSUM_LOCATION; `
+    $GotestsumBinPath=Join-Path -Path $GotestsumPath -ChildPath "gotestsum.exe"; `
+    if (Test-Path $GotestsumBinPath) { `
+      Write-Host "INFO: gotestsum already available in $cwGotestsumBinPath - skipping building it...." `
```
Done

vikramhh commented on Dockerfile.windows:
```diff
@@ -249,19 +255,87 @@ RUN `
+    } `
+    $optsForGet = @('"get"', '"-d"', '"gotest.tools/gotestsum"'); ` 
+    $logFileName = 'gotestsum.get.out';  `
+    &go $optsForGet > $logFileName 2>&1; `
```
Done

vikramhh commented on Dockerfile.windows:
```diff
@@ -249,19 +255,87 @@ RUN `
+      $savedExitCode = $LASTEXITCODE; `
+      Save-FileIfNonZero -LogName $logFileName; `
+    } else { `
+      Throw '"gotestsum checkout failed..."'; `
```
Done

vikramhh commented on Dockerfile.windows:
```diff
@@ -249,19 +255,87 @@ RUN `
+    Write-Host "INFO:     Sources obtained for gotestsum..."; `
+    Push-Location $GotestsumPath; `
+    $optsForCheckout = @('"checkout"', '"-q"', """$Env:GOTESTSUM_COMMIT"""); ` 
+    $logFileName = 'gotestsum.git.out';  `
```
I made the change so that we are not hiding the output

vikramhh commented on Dockerfile.windows:
```diff
@@ -249,19 +255,87 @@ RUN `
+    Push-Location $GotestsumPath; `
+    $optsForCheckout = @('"checkout"', '"-q"', """$Env:GOTESTSUM_COMMIT"""); ` 
+    $logFileName = 'gotestsum.git.out';  `
+    &git $optsForCheckout > $logFileName 2>&1; `
```
Same change made in all the places where we were using logs earlier. Logs are no longer being used.

vikramhh commented on Dockerfile.windows:
```diff
@@ -249,19 +255,87 @@ RUN `
+      Write-Host "INFO:     Checkout done for gotestsum..."; `
+      $logFileName = 'gotestsum.build.out';  `
+      $optsForBuild = @('"build"', '"-buildmode=exe"'); ` 
+      &go $optsForBuild > gotestsum.build.out 2>&1; `
```
Done

thaJeztah commented on hack/ci/windows.ps1:
```diff
@@ -864,10 +903,12 @@ Try {
-                if (-not($LastExitCode -eq 0)) {
+                # Copy all the test results to TEMPORIG for archival
+                Copy-Item -Path "$env:SOURCES_DRIVE`:\$env:SOURCES_SUBDIR\src\github.com\docker\docker\bundles\junit-report*xml" -Destination $TEMPORIG
+                if (-not($IntTestsRunResult -eq 0)) {
```
I realise this was already there, just curious; is there any difference between this, and

```powershell
if ($IntTestsRunResult -ne 0) {
```

vikramhh replied:
This is something we need to revisit immediately after these changes go through (for the issue where our Unix output does not match the actual results of test runs [passed or failed]). So if you feel strongly about using a particular pattern, please document it and we can use it for that change.

thaJeztah replied:
Yeah, not a big deal for now, but `-ne 0` is easier to grasp than the "double negative" used currently.

vikramhh replied:
Absolutely - will make sure to change it with that fix.

thaJeztah commented on hack/make.ps1:
```diff
@@ -319,38 +320,53 @@ Function Run-UnitTests() {
+        if ($normDir.StartsWith("-"))
+        {
+            $normDir = $normDir.TrimStart("-")
+            Write-Host $normDir
```
Think this was for debugging

vikramhh replied:
Removed

thaJeztah commented on hack/make.ps1:
```diff
@@ -319,38 +320,53 @@ Function Run-UnitTests() {
+        if ($normDir.EndsWith("-"))
+        {
+            $normDir = $normDir.TrimEnd("-")
+            Write-Host $normDir
```
Same

vikramhh replied:
Removed

thaJeztah commented on Dockerfile.windows:
```diff
@@ -249,19 +252,60 @@ RUN `
+
+  # Need a different layer  because the current layer  will not
+  # have go in its path. We could always figure that out either
+  # by observing  the current  behavior of  go-s installer (and 
```
I'm a bit confused by this, because Go doesn't have an installer; we're downloading a zip and extract it to `C:\go`. We update the path ourselves at this line; https://github.com/moby/moby/blob/c057825cf56850ffb97cae532d0bfa261b4b9a53/Dockerfile.windows#L211

Does Windows only evaluate `PATH` when the shell is started, and not when updated in the same shell?

vikramhh replied:
That is correct[confirmed the behavior from a Powershell command prompt]. The below works - or just doing it in another layer, which is what I chose to do. 

`  $env:Path += """;c:\go\bin"""; `
`  $env:Path += """;c:\git\cmd"""; `

My A/B testing did not produce a clear "winner" so it looks like we are good either way.

thaJeztah commented on Dockerfile.windows:
```diff
@@ -249,19 +252,60 @@ RUN `
+      Throw '"Failed getting gotestsum sources..."'  `
+    }; `
+    Write-Host "INFO:     Sources obtained for gotestsum..."; `
+    Push-Location $GotestsumPath; `
```
Looks like `$GotestsumPath` is only used here, so better to move declaration of that variable also here

vikramhh replied:
Done

thaJeztah commented on Dockerfile.windows:
```diff
@@ -249,19 +252,60 @@ RUN `
+    Write-Host "INFO:     Sources obtained for gotestsum..."; `
+    Push-Location $GotestsumPath; `
+    $optsForCheckout = @('"checkout"', '"-q"', """$Env:GOTESTSUM_COMMIT"""); ` 
+    &git $optsForCheckout; `
```
Still in favour of inlining these (I'll write up what I think it could look like below)

vikramhh replied:
I feel this should not be a blocker for the PR.

thaJeztah commented on Dockerfile.windows:
```diff
@@ -249,19 +252,60 @@ RUN `
+  } `
   `
-  Write-Host INFO: Completed
+  Set-Location -Path $Env:GOPATH; `
```
The function immediately changes to `$GotestsumPath`, so likely this isn't needed.

vikramhh replied:
Removed

thaJeztah commented on Dockerfile.windows:
```diff
@@ -249,19 +252,60 @@ RUN `
+    $optsForCheckout = @('"checkout"', '"-q"', """$Env:GOTESTSUM_COMMIT"""); ` 
+    &git $optsForCheckout; `
+    $savedExitCode = $LASTEXITCODE; `
+    if ($savedExitCode -eq 0) { `
```
If you reverse the logic here, we can `Throw` directly here, and probably don't need the `$savedExitCode`. With the changes I suggested, it would look like;

```powershell
  Function Build-GoTestSum() { `
    Write-Host "INFO: Building gotestsum version $Env:GOTESTSUM_COMMIT"; `
    &go get -d gotest.tools/gotestsum; `
    if ($LASTEXITCODE -ne 0) {  `
      Throw '"Failed getting gotestsum sources..."'  `
    }; `
    Write-Host "INFO:     Sources obtained for gotestsum..."; `
    $GotestsumPath=Join-Path -Path $Env:GOPATH -ChildPath "src\gotest.tools\gotestsum"; `
    Push-Location $GotestsumPath; `
    &git checkout -q "$Env:GOTESTSUM_COMMIT"; `
    if ($LASTEXITCODE -ne 0) { `
      Throw '"gotestsum checkout failed..."'; `
    } `
    Write-Host "INFO:     Checkout done for gotestsum..."; `
    &go build --buildmode=exe; `
    if ($LASTEXITCODE -ne 0) {  `
      Throw '"gotestsum build failed..."'; `
    } `
    Pop-Location; `
    Write-Host "INFO:     Build done for gotestsum..."; `
  } `
  `
  Build-GoTestSum
```


Looking further, perhaps if we change `ErrorActionPreference`

```powershell
$ErrorActionPreference = "SilentlyContinue"
```

In which case, we may not even need the manual handling of `LASTEXITCODE`. If I'm right, we could simplify the script quite a lot. Taking if further; given that everything is in a single run, we wouldn't have to worry about `Push-location` / `Pop-location`, and just change to the `$GotestsumPath` directory. At that point, the function is so minimal, I'd even consider just not creating a function for it;

```powershell
  Write-Host "INFO: Building gotestsum version $Env:GOTESTSUM_COMMIT"; `
  &go get -d gotest.tools/gotestsum; `
  Write-Host "INFO:     Sources obtained for gotestsum..."; `
  $GotestsumPath=Join-Path -Path $Env:GOPATH -ChildPath "src\gotest.tools\gotestsum"; `
  Set-Location -Path $GotestsumPath; `
  &git checkout -q "$Env:GOTESTSUM_COMMIT"; `
  Write-Host "INFO:     Checkout done for gotestsum..."; `
  &go build --buildmode=exe; `
  Write-Host "INFO:     Build done for gotestsum...";
```

(even the `INFO` logs may be a bit redundant at that point);

```powershell
  Write-Host "INFO: Building gotestsum version $Env:GOTESTSUM_COMMIT"; `
  &go get -d gotest.tools/gotestsum; `
  Set-Location "$Env:GOPATH\src\gotest.tools\gotestsum"; `
  &git checkout -q "$Env:GOTESTSUM_COMMIT"; `
  &go build --buildmode=exe; `
  Write-Host "INFO: Build done for gotestsum...";
```

vikramhh replied:
This is a very good suggestion but I feel it falls outside the scope of current work. I would feel more comfortable making such logical changes when we can thoroughly testing them. So I would suggest that we take this up through other PR-s or cleanup.

thaJeztah replied:
FWIW; looks like PowerShell 7 will support `&&` to chain commands, which could be an option in future (although likely only for recent base images, so unlikely for RS1); see https://github.com/PowerShell/PowerShell/pull/9849

vikramhh commented:
@thaJeztah - the failures on RS5 are known issues. If you are good with the changes, could we get RS1 checks as well.

andrewhsu commented:
This PR is valuable as is with RS5 junit test results in jenkins. It is important to expose the failures since the return code is always success.

I'd recommend getting this PR in as soon as possible and having a separate PR for RS1.

thaJeztah commented:
The issue we noticed when discussing, is that the naming of the tests for Windows is different than the ones for Linux, therefore the tests show in a different place; for reference, this is how the naming is set for the Linux ones for the junit.xml; https://github.com/moby/moby/blob/e7805653b8632aae4f789dbca238e25b16df964d/hack/make/.integration-test-helpers#L69-L90

vikramhh commented:
The above is true only for integration tests and could be handled by transforming the output of all tests [i.e. the xml files] by applying the same rules to each test phase [unit/integration and integration-cli] once we figure out where each result type should show up. This is not a blocker.

thaJeztah reviewed (commented):
LGTM

thaJeztah commented:
@vikramhh could you rebase this PR, because I think the failing test is fixed / skipped on master, so after rebasing, it should be green

vikramhh commented:
Rebased - the two failures are expected because only 1 of the 3 failing tests in https://github.com/moby/moby/pull/40155 has been disabled. The other two still fail [even though Jenkins does not always mark the checks as failed, something that should be partially fixed by this PR]. The third failure will be fixed by https://github.com/moby/moby/pull/40193

thaJeztah reviewed (commented):
LGTM

(rebased, to resolve merge-conflict due to Golang bump)

cpuguy83 reviewed (approved):
No idea about the powershell, but LGTM! 🐑

thaJeztah commented:
remaining failures were tests that don't work on windows, and are now skipped through https://github.com/moby/moby/pull/40199

thaJeztah commented:
rebased to run CI once more, but moving to "merge"

thaJeztah commented:
Remaining "failure" is 

```
[2019-12-13T09:38:02.934Z] === Failed
[2019-12-13T09:38:02.934Z] === FAIL: github.com/docker/docker/integration/plugin/volumes  (0.00s)
[2019-12-13T09:38:02.934Z] INFO: Windows Base image is  mcr.microsoft.com/windows/servercore:ltsc2019
[2019-12-13T09:38:02.934Z] ok  	github.com/docker/docker/integration/plugin/volumes	0.325s
```


Which will be addressed by https://github.com/moby/moby/pull/40193

thaJeztah commented:
https://github.com/moby/moby/pull/40193 was merged, so rebased

thaJeztah commented:
All green now. As a follow-up we need to rewrite the test names to match what's used on Linux, so that the Windows tests will appear separate (instead of under `github.com/....`);

Windows currently produces:

```xml
<testsuite tests="22" failures="0" time="28.840000" name="github.com/docker/docker/integration/build">
		<properties>
			<property name="go.version" value="go1.13.4 windows/amd64"></property>
		</properties>
		<testcase classname="github.com/docker/docker/integration/build" name="TestBuildWithSession" time="0.000000">
```

The same test on Linux (amd64) produces:

```xml
	<testsuite tests="32" failures="0" time="171.970000" name="amd64.integration.build">
		<properties>
			<property name="go.version" value="go1.13.4 linux/amd64"></property>
		</properties>
		<testcase classname="amd64.integration.build" name="TestBuildWithSession" time="0.000000">
			<skipped message="=== RUN   TestBuildWithSession&#xA;--- SKIP: TestBuildWithSession (0.00s)&#xA;    build_session_test.go:25: TODO: BuildKit&#xA;"></skipped>
```
